File: backend/domain/entities/analytics.py
from typing import Optional

from pydantic import BaseModel


class AnalyticsScope(BaseModel):
    """Escopo de uma análise: pessoal (group_id vazio) ou de um grupo."""

    user_id: str
    group_id: Optional[str] = None


class PeriodRange(BaseModel):
    start: str  # YYYY-MM-DD, inclusivo
    end: str  # YYYY-MM-DD, inclusivo


class AnalyticsTotals(BaseModel):
    income: float = 0.0
    expense: float = 0.0
    goal_saved: float = 0.0
    balance: float = 0.0
    savings_rate: float = 0.0  # fração (0..1) do que sobrou da receita
    transaction_count: int = 0

    @classmethod
    def build(
        cls,
        income: float,
        expense: float,
        goal_saved: float,
        transaction_count: int,
    ) -> "AnalyticsTotals":
        balance = income - expense
        rate = (balance / income) if income > 0 else 0.0
        return cls(
            income=round(income, 2),
            expense=round(expense, 2),
            goal_saved=round(goal_saved, 2),
            balance=round(balance, 2),
            savings_rate=round(rate, 4),
            transaction_count=transaction_count,
        )


class AnalyticsSummary(BaseModel):
    period: PeriodRange
    previous_period: PeriodRange
    include_drafts: bool
    current: AnalyticsTotals
    previous: AnalyticsTotals
    delta_pct: dict[str, Optional[float]]  # income, expense, goal_saved, balance


class BreakdownItem(BaseModel):
    label: str
    total: float
    count: int
    share: float = 0.0  # fração do total do tipo (0..1)


class MonthlyPoint(BaseModel):
    month: str  # YYYY-MM
    income: float = 0.0
    expense: float = 0.0
    goal_saved: float = 0.0
    balance: float = 0.0


class UpcomingDueItem(BaseModel):
    transaction_id: str
    description: Optional[str] = None
    category: str
    subcategory: Optional[str] = None
    amount: float
    due_date: str
    days_left: int
    payment_method: Optional[str] = None
    bank: Optional[str] = None

File: apps/desktop/src/types/analytics.ts
export interface AnalyticsTotals {
  income: number;
  expense: number;
  goal_saved: number;
  balance: number;
  savings_rate: number; // 0..1
  transaction_count: number;
}

export interface PeriodRange {
  start: string; // YYYY-MM-DD
  end: string; // YYYY-MM-DD
}

export interface AnalyticsDelta {
  income: number | null;
  expense: number | null;
  goal_saved: number | null;
  balance: number | null;
}

export interface AnalyticsSummary {
  period: PeriodRange;
  previous_period: PeriodRange;
  include_drafts: boolean;
  current: AnalyticsTotals;
  previous: AnalyticsTotals;
  delta_pct: AnalyticsDelta;
}

export type BreakdownDimension = "category" | "subcategory" | "payment_method" | "bank";
export type BreakdownKind = "expense" | "income";

export interface BreakdownItem {
  label: string;
  total: number;
  count: number;
  share: number; // 0..1
}

export interface Breakdown {
  dimension: BreakdownDimension;
  kind: BreakdownKind;
  total: number;
  items: BreakdownItem[];
}

export interface MonthlyPoint {
  month: string; // YYYY-MM
  income: number;
  expense: number;
  goal_saved: number;
  balance: number;
}

export interface UpcomingDueItem {
  transaction_id: string;
  description?: string | null;
  category: string;
  subcategory?: string | null;
  amount: number;
  due_date: string;
  days_left: number;
  payment_method?: string | null;
  bank?: string | null;
}

export interface UpcomingDue {
  total: number;
  items: UpcomingDueItem[];
}

export interface AnalyticsQuery {
  from: string;
  to: string;
  includeDrafts: boolean;
}
